Null Pointer Exception while Migrating v1 Repositories in class sonia.scm.update.repository.PublicFlagUpdateStep

Trying to test migration from v1.6 to v2.7 on a virtual box using debian 10.
I only copied 2 of our 100+ repos for the test. 

The migration worked once, but afterwards (same version) failed with an error message.

Any ideas?

See the part of the log here:

```
2020-10-13 21:36:36.829 [main] [          ] INFO  sonia.scm.update.UpdateEngine - running update step for type sonia.scm.repository.xml and version 2.0.3 (class sonia.scm.update.repository.PublicFlagUpdateStep)
2020-10-13 21:36:36.829 [main] [          ] INFO  sonia.scm.update.repository.PublicFlagUpdateStep - Migrating public flags of repositories as RepositoryRolePermission 'READ' for user '_anonymous'
2020-10-13 21:36:36.962 [main] [          ] INFO  sonia.scm.update.repository.PublicFlagUpdateStep - Create new _anonymous user
2020-10-13 21:36:36.963 [main] [          ] INFO  sonia.scm.update.repository.PublicFlagUpdateStep - Delete obsolete anonymous user
2020-10-13 21:36:36.969 [main] [          ] ERROR sonia.scm.migration.UpdateException - could not execute update for type sonia.scm.repository.xml to version 2.0.3 in class sonia.scm.update.repository.PublicFlagUpdateStep
java.lang.NullPointerException: null
	at sonia.scm.update.repository.PublicFlagUpdateStep.lambda$addRepositoryReadPermissionForAnonymousUser$1(PublicFlagUpdateStep.java:92)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at sonia.scm.update.repository.PublicFlagUpdateStep.addRepositoryReadPermissionForAnonymousUser(PublicFlagUpdateStep.java:90)
	at sonia.scm.update.repository.PublicFlagUpdateStep.lambda$doUpdate$0(PublicFlagUpdateStep.java:70)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at sonia.scm.update.repository.PublicFlagUpdateStep.doUpdate(PublicFlagUpdateStep.java:66)
	at sonia.scm.update.UpdateEngine.execute(UpdateEngine.java:87)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at sonia.scm.update.UpdateEngine.update(UpdateEngine.java:77)
	at sonia.scm.lifecycle.BootstrapContextListener.processUpdates(BootstrapContextListener.java:175)
	at sonia.scm.lifecycle.BootstrapContextListener.startEitherMigrationOrApplication(BootstrapContextListener.java:133)
	at sonia.scm.lifecycle.BootstrapContextListener.createMigrationOrNormalModuleProvider(BootstrapContextListener.java:124)
	at sonia.scm.lifecycle.BootstrapContextListener.createStageTwoInjector(BootstrapContextListener.java:117)
	at sonia.scm.lifecycle.BootstrapContextListener.getInjector(BootstrapContextListener.java:93)
	at com.google.inject.servlet.GuiceServletContextListener.contextInitialized(GuiceServletContextListener.java:45)
	at sonia.scm.lifecycle.BootstrapContextListener.contextInitialized(BootstrapContextListener.java:75)
	at sonia.scm.lifecycle.BootstrapContextFilter.initializeContext(BootstrapContextFilter.java:74)
	at sonia.scm.lifecycle.BootstrapContextFilter.init(BootstrapContextFilter.java:65)
	at org.eclipse.jetty.servlet.FilterHolder.initialize(FilterHolder.java:134)
	at org.eclipse.jetty.servlet.ServletHandler.lambda$initialize$0(ServletHandler.java:767)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.eclipse.jetty.servlet.ServletHandler.initialize(ServletHandler.java:791)
	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:360)
	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1445)
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1409)
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:855)
	at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:275)
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:524)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:117)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
	at org.eclipse.jetty.server.Server.start(Server.java:408)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:110)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
	at org.eclipse.jetty.server.Server.doStart(Server.java:372)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at sonia.scm.server.ScmServer.init(ScmServer.java:130)
	at sonia.scm.server.ScmServer.run(ScmServer.java:91)
	at sonia.scm.server.ScmServerDaemon.main(ScmServerDaemon.java:53)
2020-10-13 21:36:36.969 [main] [          ] ERROR sonia.scm.lifecycle.BootstrapContextListener - failed to create stage two injector
sonia.scm.migration.UpdateException: could not execute update for type sonia.scm.repository.xml to version 2.0.3 in class sonia.scm.update.repository.PublicFlagUpdateStep
	at sonia.scm.update.UpdateEngine.execute(UpdateEngine.java:90)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at sonia.scm.update.UpdateEngine.update(UpdateEngine.java:77)
	at sonia.scm.lifecycle.BootstrapContextListener.processUpdates(BootstrapContextListener.java:175)
	at sonia.scm.lifecycle.BootstrapContextListener.startEitherMigrationOrApplication(BootstrapContextListener.java:133)
	at sonia.scm.lifecycle.BootstrapContextListener.createMigrationOrNormalModuleProvider(BootstrapContextListener.java:124)
	at sonia.scm.lifecycle.BootstrapContextListener.createStageTwoInjector(BootstrapContextListener.java:117)
	at sonia.scm.lifecycle.BootstrapContextListener.getInjector(BootstrapContextListener.java:93)
	at com.google.inject.servlet.GuiceServletContextListener.contextInitialized(GuiceServletContextListener.java:45)
	at sonia.scm.lifecycle.BootstrapContextListener.contextInitialized(BootstrapContextListener.java:75)
	at sonia.scm.lifecycle.BootstrapContextFilter.initializeContext(BootstrapContextFilter.java:74)
	at sonia.scm.lifecycle.BootstrapContextFilter.init(BootstrapContextFilter.java:65)
	at org.eclipse.jetty.servlet.FilterHolder.initialize(FilterHolder.java:134)
	at org.eclipse.jetty.servlet.ServletHandler.lambda$initialize$0(ServletHandler.java:767)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.eclipse.jetty.servlet.ServletHandler.initialize(ServletHandler.java:791)
	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:360)
	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1445)
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1409)
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:855)
	at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:275)
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:524)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:117)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
	at org.eclipse.jetty.server.Server.start(Server.java:408)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:110)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
	at org.eclipse.jetty.server.Server.doStart(Server.java:372)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at sonia.scm.server.ScmServer.init(ScmServer.java:130)
	at sonia.scm.server.ScmServer.run(ScmServer.java:91)
	at sonia.scm.server.ScmServerDaemon.main(ScmServerDaemon.java:53)
Caused by: java.lang.NullPointerException: null
	at sonia.scm.update.repository.PublicFlagUpdateStep.lambda$addRepositoryReadPermissionForAnonymousUser$1(PublicFlagUpdateStep.java:92)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at sonia.scm.update.repository.PublicFlagUpdateStep.addRepositoryReadPermissionForAnonymousUser(PublicFlagUpdateStep.java:90)
	at sonia.scm.update.repository.PublicFlagUpdateStep.lambda$doUpdate$0(PublicFlagUpdateStep.java:70)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at sonia.scm.update.repository.PublicFlagUpdateStep.doUpdate(PublicFlagUpdateStep.java:66)
	at sonia.scm.update.UpdateEngine.execute(UpdateEngine.java:87)
	... 45 common frames omitted
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,10 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## Unreleased
+### Fixed
+- Null Pointer Exception on anonymous migration with deleted repositories ([#1371](https://github.com/scm-manager/scm-manager/pull/1371))
+
 ## [2.7.0] - 2020-10-12
 ### Added
 - Users can create API keys with limited permissions ([#1359](https://github.com/scm-manager/scm-manager/pull/1359))

--- a/scm-webapp/src/main/java/sonia/scm/update/repository/PublicFlagUpdateStep.java
+++ b/scm-webapp/src/main/java/sonia/scm/update/repository/PublicFlagUpdateStep.java
@@ -90,9 +90,11 @@ public class PublicFlagUpdateStep implements UpdateStep {
       .forEach(v1Repository -> {
         Repository v2Repository = repositoryDAO.get(v1Repository.getId());
         if (v2Repository != null) {
-          LOG.info(String.format("Add RepositoryRole 'READ' to _anonymous user for repository: %s - %s/%s", v2Repository.getId(), v2Repository.getNamespace(), v2Repository.getName()));
+          LOG.info("Add RepositoryRole 'READ' to _anonymous user for repository: {} - {}/{}", v2Repository.getId(), v2Repository.getNamespace(), v2Repository.getName());
           v2Repository.addPermission(new RepositoryPermission(v2AnonymousUser.getId(), "READ", false));
           repositoryDAO.modify(v2Repository);
+        } else {
+          LOG.info("Repository no longer found for id {}; could not set permission for former anonymous mode", v1Repository.getId());
         }
       });
   }

--- a/scm-webapp/src/main/java/sonia/scm/update/repository/PublicFlagUpdateStep.java
+++ b/scm-webapp/src/main/java/sonia/scm/update/repository/PublicFlagUpdateStep.java
@@ -21,7 +21,7 @@
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
-    
+
 package sonia.scm.update.repository;
 
 import org.slf4j.Logger;
@@ -89,9 +89,11 @@ public class PublicFlagUpdateStep implements UpdateStep {
       .filter(V1Repository::isPublic)
       .forEach(v1Repository -> {
         Repository v2Repository = repositoryDAO.get(v1Repository.getId());
-        LOG.info(String.format("Add RepositoryRole 'READ' to _anonymous user for repository: %s - %s/%s", v2Repository.getId(), v2Repository.getNamespace(), v2Repository.getName()));
-        v2Repository.addPermission(new RepositoryPermission(v2AnonymousUser.getId(), "READ", false));
-        repositoryDAO.modify(v2Repository);
+        if (v2Repository != null) {
+          LOG.info(String.format("Add RepositoryRole 'READ' to _anonymous user for repository: %s - %s/%s", v2Repository.getId(), v2Repository.getNamespace(), v2Repository.getName()));
+          v2Repository.addPermission(new RepositoryPermission(v2AnonymousUser.getId(), "READ", false));
+          repositoryDAO.modify(v2Repository);
+        }
       });
   }
 

--- a/scm-webapp/src/test/java/sonia/scm/update/repository/PublicFlagUpdateStepTest.java
+++ b/scm-webapp/src/test/java/sonia/scm/update/repository/PublicFlagUpdateStepTest.java
@@ -25,6 +25,7 @@
 package sonia.scm.update.repository;
 
 import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.junit.jupiter.api.io.TempDir;
@@ -77,63 +78,81 @@ class PublicFlagUpdateStepTest {
     //prepare backup xml
     V1RepositoryFileSystem.createV1Home(tempDir);
     Files.move(tempDir.resolve("config").resolve("repositories.xml"), tempDir.resolve("config").resolve("repositories.xml.v1.backup"));
-    when(repositoryDAO.get((String) any())).thenReturn(REPOSITORY);
   }
 
   @Test
-  void shouldDeleteOldAnonymousUserIfExists() throws JAXBException {
-    User anonymous = new User("anonymous");
-    when(userDAO.getAll()).thenReturn(Collections.singleton(anonymous));
-    doReturn(anonymous).when(userDAO).get("anonymous");
-    doReturn(SCMContext.ANONYMOUS).when(userDAO).get(SCMContext.USER_ANONYMOUS);
-
-    updateStep.doUpdate();
-
-    verify(userDAO).delete(anonymous);
-  }
-
-  @Test
-  void shouldNotTryToDeleteOldAnonymousUserIfNotExists() throws JAXBException {
-    when(userDAO.getAll()).thenReturn(Collections.emptyList());
-    doReturn(SCMContext.ANONYMOUS).when(userDAO).get(SCMContext.USER_ANONYMOUS);
-
-    updateStep.doUpdate();
-
-    verify(userDAO, never()).delete(any());
-  }
-
-  @Test
-  void shouldCreateNewAnonymousUserIfNotExists() throws JAXBException {
-    doReturn(SCMContext.ANONYMOUS).when(userDAO).get(SCMContext.USER_ANONYMOUS);
-    when(userDAO.getAll()).thenReturn(Collections.singleton(new User("trillian")));
-
-    updateStep.doUpdate();
-
-    verify(userDAO).add(SCMContext.ANONYMOUS);
-  }
-
-  @Test
-  void shouldNotCreateNewAnonymousUserIfAlreadyExists() throws JAXBException {
-    doReturn(SCMContext.ANONYMOUS).when(userDAO).get(SCMContext.USER_ANONYMOUS);
-    when(userDAO.getAll()).thenReturn(Collections.singleton(new User("_anonymous")));
-
-    updateStep.doUpdate();
-
-    verify(userDAO, never()).add(SCMContext.ANONYMOUS);
-  }
-
-  @Test
-  void shouldMigratePublicFlagToAnonymousRepositoryPermission() throws JAXBException {
+  void shouldNotFailForDeletedRepository() throws JAXBException {
     when(userDAO.getAll()).thenReturn(Collections.emptyList());
     when(userDAO.get("_anonymous")).thenReturn(SCMContext.ANONYMOUS);
 
     updateStep.doUpdate();
 
-    verify(repositoryDAO, times(2)).modify(repositoryCaptor.capture());
+    verify(repositoryDAO, never()).modify(any());
+  }
 
-    RepositoryPermission migratedRepositoryPermission = repositoryCaptor.getValue().getPermissions().iterator().next();
-    assertThat(migratedRepositoryPermission.getName()).isEqualTo(SCMContext.USER_ANONYMOUS);
-    assertThat(migratedRepositoryPermission.getRole()).isEqualTo("READ");
-    assertThat(migratedRepositoryPermission.isGroupPermission()).isFalse();
+  @Nested
+  class WithExistingRepository {
+
+    @BeforeEach
+    void mockRepository() {
+      when(repositoryDAO.get((String) any())).thenReturn(REPOSITORY);
+    }
+
+    @Test
+    void shouldDeleteOldAnonymousUserIfExists() throws JAXBException {
+      User anonymous = new User("anonymous");
+      when(userDAO.getAll()).thenReturn(Collections.singleton(anonymous));
+      doReturn(anonymous).when(userDAO).get("anonymous");
+      doReturn(SCMContext.ANONYMOUS).when(userDAO).get(SCMContext.USER_ANONYMOUS);
+
+      updateStep.doUpdate();
+
+      verify(userDAO).delete(anonymous);
+    }
+
+    @Test
+    void shouldNotTryToDeleteOldAnonymousUserIfNotExists() throws JAXBException {
+      when(userDAO.getAll()).thenReturn(Collections.emptyList());
+      doReturn(SCMContext.ANONYMOUS).when(userDAO).get(SCMContext.USER_ANONYMOUS);
+
+      updateStep.doUpdate();
+
+      verify(userDAO, never()).delete(any());
+    }
+
+    @Test
+    void shouldCreateNewAnonymousUserIfNotExists() throws JAXBException {
+      doReturn(SCMContext.ANONYMOUS).when(userDAO).get(SCMContext.USER_ANONYMOUS);
+      when(userDAO.getAll()).thenReturn(Collections.singleton(new User("trillian")));
+
+      updateStep.doUpdate();
+
+      verify(userDAO).add(SCMContext.ANONYMOUS);
+    }
+
+    @Test
+    void shouldNotCreateNewAnonymousUserIfAlreadyExists() throws JAXBException {
+      doReturn(SCMContext.ANONYMOUS).when(userDAO).get(SCMContext.USER_ANONYMOUS);
+      when(userDAO.getAll()).thenReturn(Collections.singleton(new User("_anonymous")));
+
+      updateStep.doUpdate();
+
+      verify(userDAO, never()).add(SCMContext.ANONYMOUS);
+    }
+
+    @Test
+    void shouldMigratePublicFlagToAnonymousRepositoryPermission() throws JAXBException {
+      when(userDAO.getAll()).thenReturn(Collections.emptyList());
+      when(userDAO.get("_anonymous")).thenReturn(SCMContext.ANONYMOUS);
+
+      updateStep.doUpdate();
+
+      verify(repositoryDAO, times(2)).modify(repositoryCaptor.capture());
+
+      RepositoryPermission migratedRepositoryPermission = repositoryCaptor.getValue().getPermissions().iterator().next();
+      assertThat(migratedRepositoryPermission.getName()).isEqualTo(SCMContext.USER_ANONYMOUS);
+      assertThat(migratedRepositoryPermission.getRole()).isEqualTo("READ");
+      assertThat(migratedRepositoryPermission.isGroupPermission()).isFalse();
+    }
   }
 }

